compiler/z/codegen/S390Peephole.cpp

```diff
@@ -1349,26 +1349,66 @@ TR_S390PostRAPeephole::ConditionalBranchReduction(TR::InstOpCode::Mnemonic branc
 
 
 /**
- * Catch the pattern where an CR/BRC can be converted
- *    CR R1,R2
- *    BRC  Mask, Lable
+ * Catch the pattern where an CLR/BRC can be converted
+ *    CLR R1, R2
+ *    BRC Mask, Lable
  * Can be replaced with
- *    CRJ  R1,R2,Lable
+ *    CLRJ R1, R2, Lable, Mask
  */
 bool
 TR_S390PostRAPeephole::CompareAndBranchReduction()
    {
-   bool performed = false;
+   if (!TR::Compiler->target.cpu.getSupportsArch(TR::CPU::z10))
+      return false;
 
-   TR::Instruction *current = _cursor;
-   TR::Instruction *next = _cursor->getNext();
-   TR::InstOpCode::Mnemonic curOpCode = current->getOpCodeValue();
-   TR::InstOpCode::Mnemonic nextOpCode = next->getOpCodeValue();
-   if ((curOpCode == TR::InstOpCode::CR || curOpCode == TR::InstOpCode::CGR || curOpCode == TR::InstOpCode::CGFR)
-      && (nextOpCode == TR::InstOpCode::BRC || nextOpCode == TR::InstOpCode::BRCL))
-      {
-      printf("Finding CR + BRC\n");
-      printf("method=%s\n", comp()->signature());
+   bool performed = false;
+   TR::Compilation *comp = _cg->comp();
+   TR::S390RRInstruction *clrInstruction = (TR::S390RRInstruction*)_cursor;
+   TR::S390BranchInstruction *brcInstruction = (TR::S390BranchInstruction*)_cursor->getNext();
+   TR::Instruction *clrjInstruction = NULL;
+
+   if (clrInstruction->getOpCodeValue() == TR::InstOpCode::CLR
+       && brcInstruction->getOpCodeValue() == TR::InstOpCode::BRC)
+      {
+      /* if branch target is a snippet we'll go ahead with the
+       * reduction. Otherwise, we look at next 5 (or until TR::BBEnd
+       * depending on whichever comes first) to check if any
+       * instruction reads or sets the condition code.
+       * We'll perform the reduction if an instruction sets
+       * the condition code or, if none of the instructions read the
+       * condition code set by clrInstruction
+       */
+      TR::LabelSymbol *labelSymbol = brcInstruction->getLabelSymbol();
+      traceMsg(comp, "\nInside CompareAndBranchReduction() for CLR and BRC");
+      if (!labelSymbol->getSnippet()) {
+         int i = 0;
+         TR::Instruction* currInstruction = brcInstruction->getNext();
+         while (currInstruction
+            && !currInstruction->getOpCode().setsCC()
+            && i < 5
+            && TR::BBEnd != currInstruction->getNode()->getOpCodeValue())
+         {
+            if (currInstruction->getOpCode().readsCC())
+               return false;
```

Comment:
I think the purpose of the `performed` variable is to have a single exit point of the function which makes debugging and understanding the implementation easier. We should either we follow that or we get rid of the `performed` variable all together.